File: LeetCode/Top Interview Questions/Array/Easy/move_zeroes_hari.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int n = nums.size();
        int i=0;
        int count = 0;
        for (i=0;i<n;i++)
        {
            if(nums[i]!=0)
            {
                swap(nums[count++],nums[i]);
                
            }
        }
    }
};

// Second attempt

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int slow=0;
        int curr=0;
        int temp;
        for(curr=0;curr<nums.size();curr++)
        {
            if(nums[curr]!=0)
            {
                temp = nums[curr];
                nums[curr]=nums[slow];
                nums[slow] = temp;
                slow++;
            }
        }
        
    }
};

File: LeetCode/LeetCode_75/move_zeroes_hari.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        slow = 0
        for i in range(len(nums)):
            if nums[i]!=0:
                temp = nums[i]
                nums[i] = nums[slow]
                nums[slow] = temp
                slow += 1








class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        slow = 0
        for i in range(len(nums)):
            if nums[i]!=0:
                nums[slow] = nums[i]
                slow += 1
        for i in range(slow,len(nums)):
            nums[i] = 0




# bad runtime , but good memory 
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        for i in nums:
            if (i==0):
                nums.remove(i)
                nums.append(i)
        return nums
    


# python optimized 

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if len(nums) < 2:
            return -1
    
        zeroes = nums.count(0) 
        
        nums[:] = [num for num in nums if num != 0]
        nums += [0] * zeroes